source-mysql/replication.go

```diff
@@ -0,0 +1,214 @@
+					Table:  table,
+				},
+			}
+
+			var columnNames = data.Table.ColumnNameString()
+			if len(columnNames) == 0 {
+				return fmt.Errorf("binlog doesn't include column names (are you running with '--binlog-row-metadata=FULL'?)")
+			}
+
+			switch event.Header.EventType {
+			case replication.WRITE_ROWS_EVENTv0, replication.WRITE_ROWS_EVENTv1, replication.WRITE_ROWS_EVENTv2:
+				for _, row := range data.Rows {
+					var after = decodeRow(columnNames, row)
+					rs.events <- sqlcapture.ChangeEvent{
+						Operation: sqlcapture.InsertOp,
+						Source:    sourceMeta,
+						After:     after,
+					}
+				}
+			case replication.UPDATE_ROWS_EVENTv0, replication.UPDATE_ROWS_EVENTv1, replication.UPDATE_ROWS_EVENTv2:
+				for rowIdx := range data.Rows {
+					// Update events contain alternating (before, after) pairs of rows
+					if rowIdx%2 == 1 {
+						var before = decodeRow(columnNames, data.Rows[rowIdx-1])
+						var after = decodeRow(columnNames, data.Rows[rowIdx])
+						rs.events <- sqlcapture.ChangeEvent{
+							Operation: sqlcapture.UpdateOp,
+							Source:    sourceMeta,
+							Before:    before,
+							After:     after,
+						}
+					}
+				}
+			case replication.DELETE_ROWS_EVENTv0, replication.DELETE_ROWS_EVENTv1, replication.DELETE_ROWS_EVENTv2:
+				for _, row := range data.Rows {
+					var before = decodeRow(columnNames, row)
+					rs.events <- sqlcapture.ChangeEvent{
+						Operation: sqlcapture.DeleteOp,
+						Source:    sourceMeta,
+						Before:    before,
+					}
+				}
+			default:
+				return fmt.Errorf("unknown row event type: %q", event.Header.EventType)
+			}
+		case *replication.XIDEvent:
+			var cursor = rs.syncer.GetNextPosition()
+			logrus.WithFields(logrus.Fields{
+				"xid":    data.XID,
+				"cursor": cursor,
+			}).Trace("XID Event")
+			rs.events <- sqlcapture.ChangeEvent{
+				Operation: sqlcapture.FlushOp,
+				Source: &mysqlSourceInfo{
+					FlushCursor: fmt.Sprintf("%s:%d", cursor.Name, cursor.Pos),
+				},
+			}
+		case *replication.TableMapEvent:
+			logrus.WithField("data", data).Trace("Table Map Event")
+		case *replication.GTIDEvent:
+			logrus.WithField("data", data).Trace("GTID Event")
+			rs.gtidTimestamp = data.OriginalCommitTime()
+		case *replication.QueryEvent:
+			logrus.WithField("data", data).Trace("Query Event")
+		default:
+			logrus.WithField("type", event.Header.EventType).Warn("unhandled event type")
+		}
+	}
+}
+
+func decodeRow(colNames []string, row []interface{}) map[string]interface{} {
+	var fields = make(map[string]interface{})
+	for idx, val := range row {
+		var name = colNames[idx]
+		if bs, ok := val.([]byte); ok {
+			val = string(bs)
+		}
+		fields[name] = val
+	}
+	return fields
+}
+
+func (rs *mysqlReplicationStream) Events() <-chan sqlcapture.ChangeEvent {
+	return rs.events
+}
+
+func (rs *mysqlReplicationStream) Acknowledge(ctx context.Context, cursor string) error {
+	// TODO(wgd): Figure out how MySQL advances whatever WAL-reserving cursor it has.
```

Comment:
This _seems_ like a question that would need to be resolved before we can really run this, unless I'm just missing some context.

Reply:
Yeah, this connector won't be ready for production usage until I finish following up on that and ideally add a test (analagous to the PostgreSQL source `TestSlotLSNAdvances`) to validate that behavior. But I'm pretty sure it won't require any fundamental changes to the overall architecture of the code, so I figured it past time to get the review started.

Reply:
ok, gotcha :+1: